lib/client/terminal/terminal_unix.go

```diff
@@ -0,0 +1,160 @@
+//go:build !windows
+// +build !windows
+
+/*
+Copyright 2021 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package terminal
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"os/signal"
+	"sync"
+	"syscall"
+
+	"github.com/gravitational/teleport/lib/utils"
+	"github.com/gravitational/trace"
+	"github.com/moby/term"
+)
+
+// Terminal is used to configure raw input and output modes for an attached
+// terminal emulator.
+type Terminal struct {
+	signalEmitter
+
+	stdin  io.Reader
+	stdout io.Writer
+	stderr io.Writer
+
+	closer    *utils.CloseBroadcaster
+	closeWait *sync.WaitGroup
+}
+
+// New creates a new Terminal instance. Callers should call `InitInteractive`
+// to configure the terminal for raw input or output modes.
+//
+// Note that the returned Terminal instance must be closed to ensure the
+// terminal is properly reset; unexpected exits may leave users' terminals
+// unusable.
+func New(stdin io.Reader, stdout, stderr io.Writer) (*Terminal, error) {
+	if stdin == nil {
+		stdin = os.Stdin
+	}
+
+	if stdout == nil {
+		stdout = os.Stdout
+	}
+
+	if stderr == nil {
+		stderr = os.Stderr
+	}
+
+	term := Terminal{
+		stdin:     stdin,
+		stdout:    stdout,
+		stderr:    stderr,
+		closer:    utils.NewCloseBroadcaster(),
+		closeWait: &sync.WaitGroup{},
+	}
+
+	return &term, nil
+}
+
+// InitInteractive puts the terminal into raw mode begins capturing raw input
+// events from the Windows API, asynchronously writing them to a Pipe emulating
+// a traditional Unix stdin.
+// Note that some implementations may replace one or more streams (particularly
+// stdin).
+func (t *Terminal) InitInteractive(input bool) error {
+	// Put the terminal into raw mode.
+	ts, err := term.SetRawTerminal(0)
+	if err != nil {
+		log.Warnf("Could not put terminal into raw mode: %v", err)
```

Comment:
This is not an error? What happens if the terminal doesn't go into raw mode?

Reply:
Without raw mode basic functionality can still (barely) work, so long as you don't try to use vim or something. On Unix in particular this was already a warning so we also don't want to add breaking behavior changes.